### DISCUSSION TITLE
Best approach to implement Rainbow Brackets

### DISCUSSION THREAD

**Discussion by Dageus** [zed-industries/zed#12159] [Feature Requests]
Hey all!,
Me and my friend @bcmeireles are trying to tackle issue #5259 and it's proving to be quite the challenge, and was wondering if someone has any idea the best approach to this.
After analyzing VSCode's implementation, it seems quite smart to use an AST, because of the fast way to recolor the brackets in case of a new addition/removal.
But Zed already seems to have a feature to highlight brackets, so i was wondering if there was a way to refactor the function that does this to have a `nesting_counter` or something like that. The only issue I'm seeing with this is the possibility that a user erases one of the brackets and the function might not be able to color the unclosed bracket correctly.
Any help is appreciated, thank you!

### useful links for reference:
 - vscode's implementation of the ast: https://github.com/microsoft/vscode/blob/eccc47493a39ad1e6ecf9b754989d3744759a10f/src/vs/editor/common/model/bracketPairsTextModelPart/bracketPairsTree/ast.ts
 - vscode's braket pairs tree: https://github.com/microsoft/vscode/blob/eccc47493a39ad1e6ecf9b754989d3744759a10f/src/vs/editor/common/model/bracketPairsTextModelPart/bracketPairsTree/bracketPairsTree.ts

**Comment by SomeoneToIgnore** [comment]
Sorry for overlooking this discussion.

As a small solace, this turned out to be quite hard in some unexpected places and no way I would be able to predict that beforehand when answering this.

PR for the reference: https://github.com/zed-industries/zed/pull/43172
